substrate/frame/staking/stake-tracker/src/lib.rs

```diff
@@ -0,0 +1,830 @@
+		Self::update_target_score(who, StakeImbalance::Negative(self_stake.into()));
+
+		// validator is a nominator too.
+		Self::on_nominator_idle(who, vec![]);
+
+		log!(debug, "on_validator_idle: {:?}, decreased self-stake {}", who, self_stake);
+	}
+
+	/// Fired when someone removes their intention to validate and has been completely removed
+	/// from the staking state.
+	///
+	/// The node is removed from the target list IFF its score is 0.
+	fn on_validator_remove(who: &T::AccountId) {
+		log!(debug, "on_validator_remove: {:?}", who,);
+
+		// validator must be idle before removing completely.
+		match T::Staking::status(who) {
+			Ok(StakerStatus::Idle) => (), // proceed
+			Ok(StakerStatus::Validator) => Self::on_validator_idle(who),
+			Ok(StakerStatus::Nominator(_)) => {
+				defensive!("on_validator_remove called on a nominator, unexpected.");
+				return
+			},
+			Err(_) => {
+				defensive!("on_validator_remove called on a non-existing target.");
+				return
+			},
+		};
+
+		// remove from target list IIF score is zero.
+		if T::TargetList::get_score(who).unwrap_or_default().is_zero() {
+			T::TargetList::on_remove(who).expect("target exists as per the check above; qed.");
+		}
+	}
+
+	/// Fired when someone sets their intention to nominate.
+	///
+	/// Note: it is assumed that `who`'s ledger staking state is updated *before* this method is
+	/// called.
+	fn on_nominator_add(who: &T::AccountId, nominations: Vec<AccountIdOf<T>>) {
+		let nominator_vote = Self::weight_of(Self::active_vote_of(who));
+
+		// voter may exist in the list in case of re-enabling a chilled nominator;
+		if T::VoterList::contains(who) {
+			return
+		}
+
+		let _ = T::VoterList::on_insert(who.clone(), nominator_vote)
+			.defensive_proof("staker does not exist in the list as per check above; qed.");
+
+		// If who is a nominator, update the vote weight of the nominations if they exist. Note:
+		// this will update the score of up to `T::MaxNominations` validators.
+		match T::Staking::status(who).defensive() {
+			Ok(StakerStatus::Nominator(_)) =>
+				for t in nominations {
+					Self::update_target_score(&t, StakeImbalance::Positive(nominator_vote.into()))
+				},
+			Ok(StakerStatus::Idle) | Ok(StakerStatus::Validator) | Err(_) => (), // nada.
+		};
+
+		log!(debug, "on_nominator_add: {:?}. role: {:?}", who, T::Staking::status(who),);
+	}
+
+	/// Fired when a nominator becomes idle (i.e. chilling).
+	///
+	/// From the `T::VotertList` PoV, chilling a nominator is the same as removing it.
+	///
+	/// Note: it is assumed that `who`'s staking ledger and `nominations` are up to date before
+	/// calling this method.
+	fn on_nominator_idle(who: &T::AccountId, nominations: Vec<T::AccountId>) {
+		Self::on_nominator_remove(who, nominations);
+	}
+
+	/// Fired when someone removes their intention to nominate and is completely removed from
+	/// the staking state.
+	///
+	/// Note: the number of nodes that are updated is bounded by the maximum number of
```

Comment:
This probably belongs in `pallet-staking` at the call site too.

Reply:
I think it's useful to keep this note for the developers who may want to use this pallet